Implicit conversion from null tensor to NoneType

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#55823 Implicit conversion from null tensor to NoneType**

Differential Revision: [D27717324](https://our.internmc.facebook.com/intern/diff/D27717324)

tugsbayasgalan committed:
Implicit conversion from null tensor to NoneType

[ghstack-poisoned]

gmagogsfm commented on test/test_jit.py:
```diff
@@ -10481,6 +10481,26 @@ def t(x):
+                super(A, self).__init__()
+
+            @torch.jit.ignore
+            def ignored(self, a: int, b: int) -> torch.Tensor:
```
Could you give me a bit more context on why it must return Tensor? I think a more pythonic way is return None
```suggestion
            def ignored(self, a: int, b: int) -> None:
```

tugsbayasgalan replied:
Sorry this was a typo

tugsbayasgalan committed:
Update on "Implicit conversion from null tensor to NoneType"


Differential Revision: [D27717324](https://our.internmc.facebook.com/intern/diff/D27717324)

[ghstack-poisoned]

driazati commented on torch/jit/annotations.py:
```diff
@@ -294,8 +290,10 @@ def is_tensor(ann):
 
 def try_ann_to_type(ann, loc):
-    if ann is None:
+    if ann is inspect.Signature.empty:
```
What's the difference between a `Signature.empty` and a `None` (i.e. how do you get one vs the other)?

tugsbayasgalan replied:
Signature.empty implies that it is coming from functions like `def func(arg1: int)` while None implies that this is coming from functions like `def func(arg1: int) -> None`. There is no real difference in python in terms of functionality

driazati commented on test/test_jit.py:
```diff
@@ -10481,6 +10481,46 @@ def t(x):
+                super(B, self).__init__()
+
+            @torch.jit.ignore
+            def ignored(self, a: int):
```
How does TS grab the return type for a script-able function these days? Does it assume it's a `Tensor` or does it try to pull out the return type from the statements in the body? This should follow the same rules as that (i.e. if no return is assumed to be `Tensor`, adding a special rule for `torch.jit.ignore` seems weird)

tugsbayasgalan replied:
I think whenever it sees the annotation is NoneType, it calls TensorType.getInferred() to annotate it as Tensor. (https://github.com/pytorch/pytorch/blob/10abbb812a47bfabba4905b50dde02a511191522/torch/jit/annotations.py#L297)

gmagogsfm commented on test/test_jit.py:
```diff
@@ -10481,6 +10481,46 @@ def t(x):
+        with self.assertRaisesRegexWithHighlight(RuntimeError, "expected value of type Tensor", "self.ignored"):
+            modelB = torch.jit.script(B())
+            modelB()
+
+
+
```
```suggestion

```

gmagogsfm commented on test/test_jit.py:
```diff
@@ -10481,6 +10481,46 @@ def t(x):
+        self.assertEqual(modelA(), 9)
+
+        with self.assertRaisesRegexWithHighlight(RuntimeError, "expected value of type Tensor", "self.ignored"):
+            modelB = torch.jit.script(B())
+            modelB()
```
Nit:

Ideally we should minimize amount of code under `assertRaises` to avoid catching unintended exceptions. Currently there are B module construction, B module instance scripting and scripted B module invocation under `assertRaises`, consider removing parts that do not trigger desired exception

gmagogsfm commented on torch/jit/annotations.py:
```diff
@@ -294,8 +290,10 @@ def is_tensor(ann):
 
 def try_ann_to_type(ann, loc):
-    if ann is None:
+    if ann is inspect.Signature.empty:
```
Fingers crossed that this doesn't break anything in sandcastle.

tugsbayasgalan committed:
Update on "Implicit conversion from null tensor to NoneType"


Differential Revision: [D27717324](https://our.internmc.facebook.com/intern/diff/D27717324)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Implicit conversion from null tensor to NoneType"


Differential Revision: [D27717324](https://our.internmc.facebook.com/intern/diff/D27717324)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Implicit conversion from null tensor to NoneType"


Differential Revision: [D27717324](https://our.internmc.facebook.com/intern/diff/D27717324)

[ghstack-poisoned]
